### PR TITLE
Resource compiler

### PR DESCRIPTION
**Pull request by Hazardu**
The resources added are defined in CMakeLists.txt inside src/irr/resources

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -35,6 +35,10 @@ if (IRR_COMPILE_WITH_CUDA)
 	enable_language(CUDA)
 endif()
 
+#resource embedding
+if(_IRR_RESOURCE_COMPILE_EVEN_IN_DEBUG)
+	add_subdirectory(src/irr/resource)
```
Comment (devshgraphicsprogramming) [inline CMakeLists.txt]: make it `src/irr/builtin`

Comment (devshgraphicsprogramming) [inline CMakeLists.txt]: make it `src/irr/builtin`

Comment (devshgraphicsprogramming) [inline CMakeLists.txt]: make it `src/irr/builtin`

File: src/irr/resource/embedResource.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    #arguments
+    name = sys.argv[1]
+    outputfilename = sys.argv[2]+ '.h'
+    inputfilename = sys.argv[3]
+
+    #opening a file
+    outp = open(outputfilename,"w+")
+
+    outp.write("#include <stdlib.h>\n")
+    outp.write("const uint8_t* %s[] = {\n" % name)
```
Comment (devshgraphicsprogramming) [inline src/irr/resource/embedResource.py]: needs to be a sized array
`const uint8_t %s[%d]` with %d giving the size

File: src/irr/resource/exampleResource.txt [hunk 3]
```diff
@@ -0,0 +1 @@
+aaaaaaaaaaaaaaaaaaaaaaaaaaab
```
Comment (devshgraphicsprogramming) [inline src/irr/resource/exampleResource.txt]: make an example out of the built-in shaders (for that you need to work off the `shader_pipeline` branch)

File: CMakeLists.txt [hunk 4]
```diff
@@ -35,6 +35,10 @@ if (IRR_COMPILE_WITH_CUDA)
 	enable_language(CUDA)
 endif()
 
+#resource embedding
+if(_IRR_RESOURCE_COMPILE_EVEN_IN_DEBUG)
```
Comment (devshgraphicsprogramming) [inline CMakeLists.txt]: you need to declare this option with a proper default, also I'd call it `IRR_EMBED_BUILTIN_RESOURCES`

it also needs to be defined in the installed configure header in `include/irr/config` so the C++ code can check the macro `_IRR_EMBED_BUILTIN_RESOURCES` and decide whether to load from embedded `uint8_t`  array or some file in `include/irr/builtin/`

**Comment by devshgraphicsprogramming** [comment]
nice stuff but needs an example of usage.

File: src/irr/builtin/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,21 @@
+cmake_minimum_required(VERSION 3.9)
```
Comment (devshgraphicsprogramming) [inline src/irr/builtin/CMakeLists.txt]: no need for this, its in the top level cmake

File: src/irr/builtin/shader1_res.h [hunk 6]
```diff
@@ -0,0 +1,41 @@
+#include <stdlib.h>
```
Comment (devshgraphicsprogramming) [inline src/irr/builtin/shader1_res.h]: don't commit the resource files.

File: src/irr/builtin/embedResource.py [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    outputfilename = inputfilename + '.h'
+    #outputfilename = re.sub(r'\..+', '', inputfilename)+ '.h'
+
+    #file size
+    # fsize = os.stat(inputfilename).st_size
+
+    #opening a file
+    outp = open(outputfilename,"w+")
+
+    outp.write("#include <stdlib.h>\n")
```
Comment (devshgraphicsprogramming) [inline src/irr/builtin/embedResource.py]: you need to embed the arrays in an appropriate namespace
```cpp
namespace irr
{
namespace builtin
{
```

File: src/irr/builtin/embedResource.py [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+
+    # outp.write("const uint8_t* resource[%d] = {\n" % fsize) # array size is not required
+    outp.write("const uint8_t* resource[] = {\n")
```
Comment (devshgraphicsprogramming) [inline src/irr/builtin/embedResource.py]: you need a unique variable name instead of `resource`

I'd take the original virtual path like `irr/builtin/X/Y/Z/something.type` and I would replace the `/` or `\` and `.` with `_` and use that as a name, for now

File: src/irr/builtin/shader1.txt [hunk 9]
```diff
@@ -0,0 +1,10 @@
+#https://github.com/Crisspl/IrrlichtBAW/blob/shaders/src/irr/asset/IAssetManager.cpp#L201
```
Comment (devshgraphicsprogramming) [inline src/irr/builtin/shader1.txt:1]: give this file the same name as its virtual path.

File: src/irr/builtin/CMakeLists.txt [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+
+function(irr_add_resource filename)
+  message(STATUS "Starting generating resource from ${filename}")
+  add_custom_command(
+    OUTPUT ${CMAKE_CURRENT_SOURCE_DIR}/${filename}.h
```
Comment (devshgraphicsprogramming) [inline src/irr/builtin/CMakeLists.txt]: you don't output to source dir, you output to binary/build dir

Comment (devshgraphicsprogramming) [reply src/irr/builtin/CMakeLists.txt]: not sure you should be using the `CMAKE_CURRENT_SOURCE` variable

File: include/irr/config/BuildConfigOptions.h.in [hunk 11]
```diff
@@ -21,5 +21,6 @@
 
 // extra config
 #cmakedefine __IRR_FAST_MATH
+#cmakedefine _IRR_EMBED_BUILTIN_RESOURCES
```
Comment (devshgraphicsprogramming) [inline include/irr/config/BuildConfigOptions.h.in]: you're missing the magical glue that defines it for this config header (should be in `./source/Irrlicht/CMakeLists.txt`)

File: src/irr/builtin/CMakeLists.txt (hunk 5 above)
Comment (devshgraphicsprogramming) [reply src/irr/builtin/CMakeLists.txt]: ?

File: src/irr/builtin/shader1_res.h (hunk 6 above)
Comment (devshgraphicsprogramming) [reply src/irr/builtin/shader1_res.h]: ?

File: src/irr/builtin/builtinDataGen.py [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+
+    #arguments
+    outputFilename = sys.argv[1]
+    cmakeSourceDir = sys.argv[2]
+    resourcePaths = sys.argv[3].split(';')
+    print(sys.argv[3])
+    #opening a file
+    outp = open(outputFilename,"w+")
+  
+    outp.write("#include irr/builtin/builtinResources.h\n\n")
```
Comment (devshgraphicsprogramming) [inline src/irr/builtin/builtinDataGen.py]: missing quotes

**Comment by devshgraphicsprogramming** [comment]
You need to compile this and test it works

File: src/irr/builtin/CMakeLists.txt [hunk 13]
```diff
@@ -2,22 +2,22 @@ function(irr_add_builtin_resource)
...
-    COMMAND ${PYTHON_EXECUTABLE} ${CMAKE_CURRENT_SOURCE_DIR}/builtinDataGen.py ${CMAKE_CURRENT_BINARY_DIR}/build/builtinResourceData.cpp ${CMAKE_SOURCE_DIR} ${BUILTIN_RESOURCE_LIST}
+    OUTPUT ${CMAKE_CURRENT_BINARY_DIR}/builtinResourceData.cpp
+    COMMAND ${PYTHON_EXECUTABLE} ${CMAKE_CURRENT_SOURCE_DIR}/builtinDataGen.py ${CMAKE_CURRENT_BINARY_DIR}/builtinResourceData.cpp ${CMAKE_SOURCE_DIR} ${BUILTIN_RESOURCE_LIST}
     COMMENT "Generated built-in resource binary data"
     VERBATIM
   )
 
   # set up dependency
-  add_custom_target(builtin_resources DEPENDS ${CMAKE_CURRENT_BINARY_DIR}/build/builtinResourceData.cpp ${CMAKE_CURRENT_BINARY_DIR}/build/builtinResources.h)
+  add_custom_target(builtin_resources DEPENDS "${CMAKE_CURRENT_BINARY_DIR}/builtinResourceData.cpp" "${CMAKE_CURRENT_BINARY_DIR}/builtinResources.h")
```
Comment (devshgraphicsprogramming) [inline src/irr/builtin/CMakeLists.txt]: check if you have the paths right, you can use cmake `message` to check what `${CMAKE_CURRENT_BINARY_DIR}/builtin` evalutates to

**Comment by devshgraphicsprogramming** [comment]
Nice stuff

File: src/irr/builtin/builtinshader.cpp [hunk 14]
```diff
@@ -0,0 +1,11 @@
+#include "irr/builtin/resuorces.h"
```
Comment (devshgraphicsprogramming) [inline src/irr/builtin/builtinshader.cpp]: don't commit these

File: src/irr/builtin/CMakeLists.txt [hunk 15]
```diff
@@ -1,19 +1,33 @@
...
+  # execute_process(COMMAND python ${CMAKE_CURRENT_SOURCE_DIR}/builtinDataGen.py ${CMAKE_CURRENT_SOURCE_DIR}/builtinResourceData.cpp ${CMAKE_SOURCE_DIR} ${BUILTIN_RESOURCE_LIST})
 endfunction()
 
 
 if(IRR_EMBED_BUILTIN_RESOURCES)
-	irr_add_resource(shader1.txt embeddedShader)
+  #fill this list with resource paths relative to top level of IrrlichtBaW's source
+  set(BUILTIN_RESOURCE_LIST src/irr/builtin/shader1.txt)
+
+	irr_add_builtin_resource()
```
Comment (devshgraphicsprogramming) [inline src/irr/builtin/CMakeLists.txt:22]: this has really bad syntax

File: src/irr/builtin/CMakeLists.txt [hunk 16]
```diff
@@ -1,19 +1,33 @@
...
+  # for testing purposes
+  # execute_process(COMMAND python ${CMAKE_CURRENT_SOURCE_DIR}/builtinHeaderGen.py ${CMAKE_CURRENT_SOURCE_DIR}/builtinResources.h ${CMAKE_SOURCE_DIR} ${BUILTIN_RESOURCE_LIST})
+  # execute_process(COMMAND python ${CMAKE_CURRENT_SOURCE_DIR}/builtinDataGen.py ${CMAKE_CURRENT_SOURCE_DIR}/builtinResourceData.cpp ${CMAKE_SOURCE_DIR} ${BUILTIN_RESOURCE_LIST})
 endfunction()
 
 
 if(IRR_EMBED_BUILTIN_RESOURCES)
-	irr_add_resource(shader1.txt embeddedShader)
+  #fill this list with resource paths relative to top level of IrrlichtBaW's source
+  set(BUILTIN_RESOURCE_LIST src/irr/builtin/shader1.txt)
```
Comment (devshgraphicsprogramming) [inline src/irr/builtin/CMakeLists.txt:20]: why is `BUILTIN_RESOURCE_LIST` some sort of global variable?

File: src/irr/builtin/CMakeLists.txt [hunk 17]
```diff
@@ -2,19 +2,17 @@ function(irr_add_builtin_resource)
 
   # generate header file
   add_custom_command(
-    OUTPUT "${CMAKE_CURRENT_BINARY_DIR}/build/builtinResources.h" "${CMAKE_CURRENT_BINARY_DIR}/build/builtinResourceData.cpp"
-    COMMAND python ${CMAKE_CURRENT_SOURCE_DIR}/buitlinHeaderGen.py ${CMAKE_CURRENT_BINARY_DIR}/build/builtinResources.h ${CMAKE_SOURCE_DIR} ${BUILTIN_RESOURCE_LIST}
-    COMMAND ${PYTHON_EXECUTABLE} ${CMAKE_CURRENT_SOURCE_DIR}/builtinDataGen.py ${CMAKE_CURRENT_BINARY_DIR}/build/builtinResourceData.cpp ${CMAKE_SOURCE_DIR} ${BUILTIN_RESOURCE_LIST}
+    OUTPUT "${CMAKE_CURRENT_BUILD_DIR}/builtinResources.h" "${CMAKE_CURRENT_BUILD_DIR}/builtinResourceData.cpp"
+    COMMAND python ${CMAKE_CURRENT_SOURCE_DIR}/builtinHeaderGen.py ${CMAKE_CURRENT_BUILD_DIR}/builtinResources.h ${CMAKE_SOURCE_DIR} ${BUILTIN_RESOURCE_LIST}
```
Comment (devshgraphicsprogramming) [inline src/irr/builtin/CMakeLists.txt]: `${PYTHON_EXECUTABLE}` is the correct one to use

File: src/irr/builtin/CMakeLists.txt [hunk 18]
```diff
@@ -2,15 +2,15 @@ function(irr_add_builtin_resource)
 
   # generate header file
   add_custom_command(
-    OUTPUT "${CMAKE_CURRENT_BUILD_DIR}/builtinResources.h" "${CMAKE_CURRENT_BUILD_DIR}/builtinResourceData.cpp"
-    COMMAND python ${CMAKE_CURRENT_SOURCE_DIR}/builtinHeaderGen.py ${CMAKE_CURRENT_BUILD_DIR}/builtinResources.h ${CMAKE_SOURCE_DIR} ${BUILTIN_RESOURCE_LIST}
-    COMMAND ${PYTHON_EXECUTABLE} ${CMAKE_CURRENT_SOURCE_DIR}/builtinDataGen.py ${CMAKE_CURRENT_BUILD_DIR}/builtinResourceData.cpp ${CMAKE_SOURCE_DIR} ${BUILTIN_RESOURCE_LIST}
+    OUTPUT "${CMAKE_CURRENT_SOURCE_DIR}/builtinResources.h" "${CMAKE_CURRENT_SOURCE_DIR}/builtinResourceData.cpp"
```
Comment (devshgraphicsprogramming) [inline src/irr/builtin/CMakeLists.txt]: WHY!?

File: src/irr/builtin/CMakeLists.txt (hunk 15 above)
Comment (devshgraphicsprogramming) [reply src/irr/builtin/CMakeLists.txt:22]: you should be passing the list to the function, not using a weird global cmake variable

**Comment by devshgraphicsprogramming** [comment]
Ok, now make sure new files are only generated in the binary/build directory (dont want changes in source dir).

You will then need to add the header to the install target (so that it get copied properly) etc.
See how we treat `BuildConfigOptions.h` which is another dynamically generated header.

Take the `src/irr/builtin/shader1.txt` and turn it into `include/irr/builtin/shader1.txt`, this is because that if we don't choose to embed the resources, they will need to be added to the INSTALL target.
Also make the `include` be implicit in the python script.

Finally instead of having `src/irr/builtin/shader1.txt` as the test case (remove it), take some of the already existing C++11 rawstring shaders in builtin loaders (such as MTL or some in IAssetManager.cpp) and move them to their `.glsl/.frag/.vert/.comp` respective files in `irr/builtin/` (have the builtin path be the same as the path for cache insertion - remember the `include` is implicit!).

**Comment by devshgraphicsprogramming** [comment]
My old example for loading shaders
```cpp
// the default but can be changed at runtime if someone needs for a reason
io::path IFileSystem::relativeIrrlichtInstallDir = "include";

// this can be an IFileSystem inline method
template<typename UniqueStringType>
inline core::smart_refctd_ptr<asset::ICPUBuffer> IFileSystem::loadBuiltinData()
{
#ifdef _IRR_EMBED_RESOURCES
  auto found = builtin::get_resource<UniqueStringType>();
  if (found.first && found.second)
    return core::make_smart_refctd_ptr<asset::CCustomAllocatorCPUBuffer<core::null_allocator<uint8_t> > >(found->second,found->first,core::adopt_memory);
  return nullptr;
#endif
  io::IFileRead* file = this->createAndOpenFile((relativeIrrlichtInstallDir+"/"+path).c_str());
  auto retval = core::make_smart_refctd_ptr<asset::ICPUBuffer>(file->getSize(),nullptr);
  file->read(retval->getPointer(),file->getSize());
  file->drop();
  
  return retval;
}
```

File: include/CConcurrentObjectCache.h [hunk 19]
```diff
@@ -4,184 +4,184 @@
 #include "CObjectCache.h"
```
Comment (devshgraphicsprogramming) [inline include/CConcurrentObjectCache.h:4]: can you revert the change to this file?

File: src/irr/asset/IAssetManager.cpp [hunk 20]
```diff
@@ -64,18 +64,9 @@
...
-#endif
-
-#ifdef _IRR_COMPILE_WITH_GLI_WRITER_
-#include "irr/asset/CGLIWriter.h"
+#ifdef _IRR_EMBED_BUILTIN_RSOURCES_
+#include "irr/builtin/builtinResources.h"
 #endif
-
-#include "irr/asset/CGLSLLoader.h"
-#include "irr/asset/CSPVLoader.h"
```
Comment (devshgraphicsprogramming) [inline src/irr/asset/IAssetManager.cpp]: why did you remove the extra headers?

File: examples_tests/03.GPU_Mesh/main.cpp [hunk 21]
```diff
@@ -4,15 +4,35 @@
 #include <irrlicht.h>
 
 //! I advise to check out this file, its a basic input handler
-#include "../common/QToQuitEventReceiver.h"
-
-//#include "../../ext/ScreenShot/ScreenShot.h"
-
+#include "../../ext/ScreenShot/ScreenShot.h"
+#include "../../src/irr/builtin/builtinResources.h"
```
Comment (devshgraphicsprogramming) [inline examples_tests/03.GPU_Mesh/main.cpp]: don't touch my samples, try to embed one of the shaders in `IAssetManager` instead of having it as a rawstring and test (example 35 uses builtin shaders)

File: include/irr/config/BuildConfigOptions.h.in [hunk 22]
```diff
@@ -18,12 +18,30 @@
 #cmakedefine _IRR_COMPILE_WITH_OPENGL_
 #cmakedefine _IRR_COMPILE_WITH_VULKAN_
 
-// compute interop
```
Comment (devshgraphicsprogramming) [inline include/irr/config/BuildConfigOptions.h.in]: WHAT THE FUCK ARE YOU DOING?

**Comment by Hazardu** [comment]
I will create a different branch with just intended changes
